--- tools/autoconf.c.orig	2012-07-03 22:55:46.955338792 +0400
+++ tools/autoconf.c	2012-07-03 22:56:17.977337001 +0400
@@ -637,7 +637,7 @@
     fprintf(makefile,"TARGET	= psybnc\n");
     fprintf(makefile,"\n");
     fprintf(makefile,"all:	$(OBJS)\n");
-    fprintf(makefile,"	$(CC) -o $(TARGET) $(CFLAGS) $(OBJS) %s $(LIBS)\n", dnsobj);
+    fprintf(makefile,"	$(CC) -o $(TARGET) $(CFLAGS) ${LDFLAGS} $(OBJS) %s $(LIBS)\n", dnsobj);
 /*    fprintf(makefile,"	@strip $(TARGET)\n");*/
     if(ssl==0)
     {
