--- configure.ac.org	2011-06-03 19:44:45.848405096 +0200
+++ configure.ac	2011-06-03 19:50:54.353341075 +0200
@@ -136,7 +136,7 @@
 PKG_CHECK_MODULES(COMPIZMOUSEPOLL, compiz-mousepoll, [have_compiz_mousepoll=yes], [have_compiz_mousepoll=no])
 AM_CONDITIONAL(SHOWMOUSE_PLUGIN, test "x$have_compiz_mousepoll" = "xyes")
 
-PKG_CHECK_MODULES(LIBNOTIFY, libnotify gtk+-2.0, [have_libnotify=yes], [have_libnotify=no])
+AC_ARG_WITH([libnotify], [AS_HELP_STRING([--with-libnotify], [Build with libnotify support @<:@default=auto@:>@])])
 AM_CONDITIONAL(NOTIFICATION_PLUGIN, test "x$have_libnotify" = "xyes")
 if test "$have_libnotify" = yes; then
   AC_DEFINE(USE_NOTIFICATION, 1, [Build notification plugin])
