--- a/Makefile.am	2013-09-25 17:35:08.835773000 +0400
+++ b/Makefile.am	2013-09-25 17:35:34.407773000 +0400
@@ -9,15 +9,12 @@
 			src/config-h.in config/stamp-h.in
 
 if EXAMPLES
-SUBDIRS = src tests docs man win32 examples
+SUBDIRS = src docs man win32 examples
 else
-SUBDIRS = src tests docs man win32
+SUBDIRS = src docs man win32
 endif
 
-docsdir = $(datadir)/wsdlpull
-docs_DATA = wsdlpull.html README
-
-EXTRA_DIST = AUTHORS COPYING CHANGES $(docs_DATA)
+EXTRA_DIST = AUTHORS COPYING CHANGES README
 
 #pkginclude_HEADERS = src/config.h
 pkgincludedir= $(includedir)

--- a/src/xmlpull/Makefile.am	2021-04-28 10:58:15.452021204 +0300
+++ b/src/xmlpull/Makefile.am	2021-04-28 10:58:25.003020502 +0300
@@ -13,5 +13,5 @@
 libwsdlpullxml_la_LDFLAGS = -lpthread -version-info $(GENERIC_LIBRARY_VERSION)
 
 pkginclude_HEADERS = $(xmlpull_h_sources)
-pkgincludedir= $(includedir)/wsdlpull/xml
+pkgincludedir= $(includedir)/wsdlpull/xmlpull
 
--- a/src/wsdlparser/Makefile.am	2021-04-28 11:00:45.428010184 +0300
+++ b/src/wsdlparser/Makefile.am	2021-04-28 11:00:55.995009408 +0300
@@ -15,5 +15,5 @@
 	-version-info $(GENERIC_LIBRARY_VERSION)
 
 # Header files to be exported
-pkginclude_HEADERS = $(schema_h_sources)
+pkginclude_HEADERS = $(schema_h_sources) $(wsdl_cc_sources)
 pkgincludedir= $(includedir)/wsdlpull/wsdlparser
