--- CMakeLists.txt.orig	2008-11-05 10:40:46.000000000 +0300
+++ CMakeLists.txt	2011-07-13 10:26:13.000000000 +0400
@@ -1,12 +1,16 @@
 project(plasma-kbstate)
 
 find_package(X11 REQUIRED)
+find_package(KDE4 REQUIRED)
+find_package(Qt4 REQUIRED)
+
+include (${QT_USE_FILE})
 
 include_directories(X11_INCLUDE_DIR)
 
 set(kbstate_SRCS kbstateapplet.cpp)
 kde4_add_plugin(plasma_applet_kbstate ${kbstate_SRCS})
-target_link_libraries(plasma_applet_kbstate ${KDE4_PLASMA_LIBS} ${KDE4_KDEUI_LIBS} ${X11_LIBRARIES})
+target_link_libraries(plasma_applet_kbstate ${KDE4_PLASMA_LIBS} ${KDE4_KDEUI_LIBS} ${X11_LIBRARIES} ${QT_LIBRARIES})
 install(TARGETS plasma_applet_kbstate DESTINATION ${PLUGIN_INSTALL_DIR})
 install(FILES plasma-applet-kbstate.desktop DESTINATION ${SERVICES_INSTALL_DIR})
 
