--- a/doc/Makefile.am	2016-02-24 16:29:44.064522852 +0300
+++ b/doc/Makefile.am	2016-02-24 16:29:12.720524705 +0300
@@ -1662,9 +1662,9 @@
 		convert $$f "${builddir}/mininim-figures/"`basename $${f%.png}.eps`; done
 
 install-info-local:
-	$(MKDIR_P) "${infodir}"
-	chmod u+w "${infodir}"
-	cp -rv "${srcdir}/mininim-figures" "${infodir}"
+	$(MKDIR_P) "$(DESTDIR)${infodir}"
+	chmod u+w "$(DESTDIR)${infodir}"
+	cp -rv "${srcdir}/mininim-figures" "$(DESTDIR)${infodir}"
 
 uninstall-local:
 	find "${infodir}/mininim-figures" \( -type f -or -type d \) \
--- a/Makefile.am	2016-02-24 16:33:45.983508550 +0300
+++ b/Makefile.am	2016-02-24 16:34:37.934505478 +0300
@@ -2105,9 +2105,9 @@
 	$(WINDRES) $(WINDRESFLAGS) -o $@ $<
 
 install-data-hook:
-	$(MKDIR_P) "${pkgdatadir}"
-	chmod u+w "${pkgdatadir}"
-	cp -rv "${top_srcdir}/data" "${pkgdatadir}"
+	$(MKDIR_P) "$(DESTDIR)${pkgdatadir}"
+	chmod u+w "$(DESTDIR)${pkgdatadir}"
+	cp -rv "${top_srcdir}/data" "$(DESTDIR)${pkgdatadir}"
 
 uninstall-local:
 	find "${pkgdatadir}/data" \( -type f -or -type d \) \
